File: backend/app/v1/schemas/timeline.py
from datetime import datetime
from typing import Any, Optional
from pydantic import BaseModel, Field
import uuid

class AssociateMarkEntry(BaseModel):
    """A single associate's weighted evaluation result for a stage."""
    associate_name: str
    marks: Optional[float] = None
    result: Optional[str] = None


class TimelineEvent(BaseModel):
    event_type: str  # "stage" or "decision"
    event_date: Optional[datetime] = None
    title: str
    description: Optional[str] = None
    result: Optional[str] = None
    ai_result: Optional[str] = None
    hr_decision: Optional[str] = None
    score: Optional[float] = None
    ai_score: Optional[float] = None
    hr_score: Optional[float] = None
    stage_id: Optional[uuid.UUID] = None
    stage_name: Optional[str] = None
    job_id: Optional[uuid.UUID] = None
    job_stage_config_id: Optional[uuid.UUID] = None
    metadata: Optional[dict[str, Any]] = None
    associate_marks: list[AssociateMarkEntry] = Field(
        default_factory=list,
        description="Associate evaluation marks (name: marks out of 5) for github+question round stages",
    )

class HiringTimelineResponse(BaseModel):
    candidate_id: uuid.UUID
    latest_decision: str = "Pending"
    current_stage: str = "Resume Screening"
    events: list[TimelineEvent]

File: frontend/src/types/timeline.ts
/**
 * Type definitions for candidate hiring timelines.
 * Matches backend schemas in timeline.py.
 */

export interface AssociateMarkEntry {
  /** A single associate's weighted evaluation result for a stage. */
  associate_name: string;
  marks?: number | null;
}

export interface TimelineEvent {
  /** Event type: 'stage' or 'decision' */
  event_type: string;
  event_date?: string | null;
  title: string;
  description?: string | null;
  result?: string | null;
  ai_result?: string | null;
  hr_decision?: string | null;
  score?: number | null;
  ai_score?: number | null;
  hr_score?: number | null;
  stage_id?: string | null;
  stage_name?: string | null;
  job_id?: string | null;
  job_stage_config_id?: string | null;
  metadata?: Record<string, any> | null;
  /** Associate evaluation marks (name: marks out of 5) for github+question round stages */
  associate_marks: AssociateMarkEntry[];
}

export interface HiringTimelineResponse {
  candidate_id: string;
  latest_decision: string;
  current_stage: string;
  events: TimelineEvent[];
}
